examples/cutting_stock_problem.rs

```diff
@@ -0,0 +1,179 @@
+///
+/// ```
+///   \max \sum_i b_i π_i
+///     s.t. \sum_i π_i z_{i,t} <= 1    [λ_t]
+///          π_i >= 0
+/// ```
+///
+/// Note that the constraint corresponding to `λ_t` is violated iff its slack, known as its reduced cost
+/// `\bar{c}(λ_t) = 1 - \sum_i π_i z_{i,t}`, is negative. Of course, instead of checking for all `t`, we introduce the pricing problem
+///
+/// ```
+///   \min 1 - \sum_i π_i z_{i}
+///     s.t. \sum_i w_i z_{i} <= W
+///          z_{i} >= 0
+/// ```
+///
+// Following https://scipbook.readthedocs.io/en/latest/bpp.html and https://www.gerad.ca/fr/papers/G-2024-36.pdf
+fn main() {
+    let stock_length = 9;
+    let item_sizes = &[6.0, 5.0, 4.0, 2.0, 3.0, 7.0, 5.0, 8.0, 4.0, 5.0];
+    let demand = &[2, 3, 4, 4, 2, 2, 2, 2, 2, 1];
+
+    // Vector of cutting_patterns, initially populated with the trivial ones that contain exactly
+    // one item. cutting_patterns[i][j] indicates how often item j is in pattern i.
+    let initial_cutting_patterns: Vec<Vec<i32>> = (0..item_sizes.len()).map(|i|
+        (0..item_sizes.len()).map(|j| if i == j { 1 } else { 0 }).collect::<Vec<i32>>()
+    ).collect();
+
+    let mut main_problem = Model::default().minimize();
+
+    let cutting_pattern_vars: Vec<Variable> = initial_cutting_patterns.iter().enumerate().map(|(i, &ref _pattern)| {
+        let pattern = (0..10).map(|x| if x == i { "1" } else { "0" }).collect::<Vec<_>>().join("-");
+        main_problem.add_var(0.0, f64::INFINITY, 1.0, &format!("pattern_{pattern}"), VarType::Integer)
+    }).collect();
+
+    demand.iter().enumerate().for_each(|(i, &count)| {
+        let cons = main_problem.add_cons(
+            cutting_pattern_vars.iter().collect(),
+            &initial_cutting_patterns.iter().map(|pattern| pattern[i] as f64).collect::<Vec<f64>>(),
+            count as f64,
+            f64::INFINITY,
+            &format!("demand_for_item_{i}"));
+        main_problem.set_cons_modifiable(&cons, true);
+    });
+
+    let pricer = CSPPricer {
+        stock_length,
+        item_sizes,
+    };
+
+    main_problem.include_pricer("CSPPricer", "CSPPricer", 1, false, Box::new(pricer));
+    let solved_model = main_problem.solve();
+
+    println!("\nSolution");
+    let solution = solved_model.best_sol().unwrap();
+    for var in solved_model.vars().iter() {
+        let name = var.name();
+        let value = solution.val(var);
+        if value != 0.0 {
+            println!("  {name}={value}")
+        }
+    }
+}
+
+struct CSPPricer<'a> {
+    stock_length: usize,
+    item_sizes: &'a [f64],
+}
+
+impl Pricer for CSPPricer<'_> {
+    fn generate_columns(&mut self, mut model: Model<Solving>, _pricer: SCIPPricer, farkas: bool) -> PricerResult {
+
+        // If the current LP relaxation is infeasible, it is the task of the pricer to generate additional variables that can potentially
+        // render the LP feasible again. This is beyond the scope of this example.
+        assert!(!farkas, "Farkas pricing not supported.");
```

Comment:
It would probably be a good idea to support this. Should not be too much trouble, in farkas pricing the objective function coefficient of the variables is then zero, and instead of the normal duals you can use the farkas duals and the rest of the code remains the same. But I could also add that in a future MR :)

Reply:
I've tried triggering the assertion unsuccessfully. I think this is because I just add the single-item-patterns as valid patterns without verifying that every item fits on the roll - making the relaxed problem feasible from the get-go.

Do you think adding unreachable code makes sense here?

Reply:
Yes it does, but only if this is after the node check it should be unreachable yes. Otherwise, infeasibility could happen deep down in the tree.